File: number.cpp
#include <iostream>
#include <ctime>
#include <cstdlib>

using namespace std;

int main(int argc, char** argv) {
    srand(time(nullptr));
    int corr = rand() % 1000 + 1;
    int guesses = 12;
    bool done = false;
    while ((!done) && (guesses > 0)) {
        cout << "Guess a number: " << endl;
        int guess;
        cin >> guess;
        if ((guess > 0) && (guess <= 1000)) {
            if (guess > corr) {
                cout << "Too high..." << endl;
            } else if (guess < corr) {
                cout << "Too low..." << endl;
            } else {
                cout << "You got it!" << endl;
                done = true;
            }
            guesses--;
        }
    }
    if (!done)
        cout << "You lose... it was " << corr << endl;
}

File: number.hs
import Data.Function(fix)
import Control.Monad
import Text.Read(readMaybe)
import System.Random

numberGame :: (Show a, Read a, Ord a) => (a, a) -> Int -> a -> IO ()
numberGame _ 0 corr = putStrLn $ "You lose. It was " ++ show corr
numberGame (mn, mx) n corr = do
  guess <- fix $ \g -> do
             putStr "> "
             gss <- readMaybe <$> getLine
             case gss of
               Nothing -> g
               Just x
                 | x >= mn && x <= mx -> return x
                 | otherwise -> g
  case compare guess corr of
    LT -> putStrLn "Too low..." >> numberGame (mn, mx) (n - 1) corr
    GT -> putStrLn "Too high..." >> numberGame (mn, mx) (n - 1) corr
    EQ -> putStrLn "You got it!"

numberGame' :: (Show a, Read a, Ord a, Random a) => (a, a) -> Int -> IO ()
numberGame' rng n = randomRIO rng >>= numberGame rng n
